core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracTombstoneManager.java

```diff
@@ -122,21 +136,29 @@ public void startCleanupTombstone() {
    }
 
    private CompletionStage<Void> performCleanup() {
-      AggregateCompletionStage<Void> stage = CompletionStages.aggregateCompletionStage();
-      for (Map.Entry<Object, TombstoneData> entry : tombstoneMap.entrySet()) {
-         DistributionInfo info = distributionManager.getCacheTopology().getSegmentDistribution(entry.getValue().getSegment());
-         if (!info.isWriteOwner()) {
-            // topology changed, no longer an owner
-            remove(entry.getKey(), entry.getValue());
-            continue;
-         }
-         if (!info.isPrimary() || iracManager.running().containsKey(entry.getKey())) {
-            // backup owner or the irac manager haven't sent the update successfully
-            continue;
+      try {
+         AggregateCompletionStage<Void> stage = CompletionStages.aggregateCompletionStage();
+         for (Map.Entry<Object, TombstoneData> entry : tombstoneMap.entrySet()) {
+            DistributionInfo info = distributionManager.getCacheTopology().getSegmentDistribution(entry.getValue().getSegment());
+            if (!info.isWriteOwner()) {
+               // topology changed, no longer an owner
+               remove(entry.getKey(), entry.getValue());
+               continue;
+            }
+            if (!info.isPrimary() || iracManager.running().containsKey(entry.getKey())) {
+               // backup owner or the irac manager haven't sent the update successfully
+               continue;
+            }
+            stage.dependsOn(new CleanupTask(entry.getKey(), entry.getValue()).checkRemoteSites());
          }
-         stage.dependsOn(new CleanupTask(entry.getKey(), entry.getValue()).checkRemoteSites());
+
+         return stage.freeze().whenComplete((unused, throwable) -> {
+            double tombstoneRatio = tombstoneMap.size() * 1.0 / configuration.tombstoneMapSize();
```

Comment:
ok, it makes sense. I tried to keep it simpler but computing the creating rate will help to achieve a better interval. Hopefully, the delay will stabilize after a while (assuming the workload does not change).

I'm also going to drop the "DynamicTask" class since everything is calculated here.